### PR TITLE
supervisor: make rejection periods work with stopTasksCount

### PR DESCRIPTION
**Pull request by adithyachakilam**
### Description

In cases where `stopTasksCount` is configured which makes streaming tasks run much longer (in some cases) than the configured task duration, early/late rejection periods are not correctly implemented since they are configured during the task startup time in IOConfig. Instead, this PR propagates updates the min/max times once every task duration.

<hr>

##### Key changed/added classes in this PR
 * `SeekableStreamIndexTaskRunner`
 * `SeekableStreamIndexTaskIOConfig`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: extensions-contrib/rabbit-stream-indexing-service/src/main/java/org/apache/druid/indexing/rabbitstream/RabbitStreamIndexTaskIOConfig.java [hunk 1]
```diff
@@ -50,8 +50,8 @@ public RabbitStreamIndexTaskIOConfig(
       @JsonProperty("consumerProperties") Map<String, Object> consumerProperties,
       @JsonProperty("pollTimeout") Long pollTimeout,
       @JsonProperty("useTransaction") Boolean useTransaction,
-      @JsonProperty("minimumMessageTime") DateTime minimumMessageTime,
-      @JsonProperty("maximumMessageTime") DateTime maximumMessageTime,
+      @JsonProperty("lateMessageRejectionPeriod") Duration lateMessageRejectionPeriod,
+      @JsonProperty("earlyMessageRejectionPeriod") Duration earlyMessageRejectionPeriod,
```
Comment (suneet-s) [inline extensions-contrib/rabbit-stream-indexing-service/src/main/java/org/apache/druid/indexing/rabbitstream/RabbitStreamIndexTaskIOConfig.java:54]: These new fields need to be in addition to the old fields to ensure a newer version of the overlord works with older versions of the indexing tasks. But more importantly, in the case of a rollback - an older overlord needs to work with a newer version of indexing services

**Review by suneet-s — COMMENTED** [review]
This implementation of refreshing the min / max times every 10 min is a change in behavior, so I think the person submitting the supervisor spec needs to opt in to this new behavior which doesn't appear possible in the current approach. 

An alternative approach to consider is to update the IOConfig to include a "refresh interval" field that defaults to infinity that would refresh the min / max times in the IOConfig. 

The tasks could then update their min / max times every duration just by adding the refresh interval so we don't need to worry about clock skew across the peons.

Since this patch is meant to just deal with stopTaskCount - have you considered refreshing the min / max times every task duration to mimic the existing behavior if stopTaskCount was not set (aka every task duration the min/max times across all peons are updated)

File: extensions-contrib/rabbit-stream-indexing-service/src/main/java/org/apache/druid/indexing/rabbitstream/RabbitStreamIndexTaskIOConfig.java (hunk 1 above)
Comment (adithyachakilam) [reply extensions-contrib/rabbit-stream-indexing-service/src/main/java/org/apache/druid/indexing/rabbitstream/RabbitStreamIndexTaskIOConfig.java:54]: May to avoid all these issues, How about instead of propagating periods, Is it okay if we just recalculate the period min/max time (and avoid introducing these new fields)?

**Comment by adithyachakilam** [comment]
@suneet-s  Thanks for the review, I've changed the patch to just update the min/max times once every task duration. So this also avoids all those version problems.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 2]
```diff
@@ -2092,4 +2109,47 @@ protected abstract void possiblyResetDataSourceMetadata(
   protected abstract boolean isEndOffsetExclusive();
 
   protected abstract TypeReference<List<SequenceMetadata<PartitionIdType, SequenceOffsetType>>> getSequenceMetadataTypeReference();
+
+
+  private void addTaskDurationToMinMaxTimes()
+  {
+    if (minMessageTime.isPresent()) {
+      minMessageTime = Optional.of(minMessageTime.get()
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:2117]: Is it ok that intelliJ is warning that this is an unsafe update of a volatile variable?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 3]
```diff
@@ -2092,4 +2109,47 @@ protected abstract void possiblyResetDataSourceMetadata(
...
+                                             minMessageTime.get().isAfter(row.getTimestamp());
+    final boolean afterMaximumMessageTime = maxMessageTime.isPresent() &&
+                                            maxMessageTime.get().isBefore(row.getTimestamp());
+
+    if (log.isDebugEnabled()) {
+      if (beforeMinimumMessageTime) {
+        log.debug(
+            "CurrentTimeStamp[%s] is before MinimumMessageTime[%s]",
+            row.getTimestamp(),
+            ioConfig.getMinimumMessageTime().get()
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:2143]:
```suggestion
            minMessageTime
```

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 4]
```diff
@@ -2092,4 +2109,47 @@ protected abstract void possiblyResetDataSourceMetadata(
...
+        );
+      } else if (afterMaximumMessageTime) {
+        log.debug(
+            "CurrentTimeStamp[%s] is after MaximumMessageTime[%s]",
+            row.getTimestamp(),
+            ioConfig.getMaximumMessageTime().get()
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:2149]:
```suggestion
            maxMessageTime
```

**Review by suneet-s — COMMENTED** [review]
I have not reviewed the tests yet.

For this change, we should also include docs for what this new parameter is in the IOConfig

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 5]
```diff
@@ -246,6 +248,9 @@ public enum Status
 
   private final Map<PartitionIdType, Long> partitionsThroughput = new HashMap<>();
 
+  private volatile Optional<DateTime> minMessageTime;
+  private volatile Optional<DateTime> maxMessageTime;
```
[the remark this answers is not in the document]
Comment (adithyachakilam) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:252]: just realized that it wouldn't overflow

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskIOConfig.java [hunk 6]
```diff
@@ -51,7 +53,8 @@ public SeekableStreamIndexTaskIOConfig(
       final Boolean useTransaction,
       final DateTime minimumMessageTime,
       final DateTime maximumMessageTime,
-      @Nullable final InputFormat inputFormat
+      @Nullable final InputFormat inputFormat,
+      @Nullable final Duration taskDuration // can be null for backward compabitility
```
[the remark this answers is not in the document]
Comment (adithyachakilam) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskIOConfig.java:57]: I think its good to have Duration and no need to convert back and forth and also confusion of adding `mins` to the variable. You think there are any issues with the schema deserializing ? (I see at least `Period` used, maybe we can switch to it if so )

Also, regarding variable name change, Do you think there would be a usecase of user editing it to a different value ? I would prefer to use `taskDuration` so that no one touches and would also avoid the need for docs (Also, I dont find this documented anywhere, can you point me to the file ?)

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java (hunk 2 above)
Comment (adithyachakilam) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:2117]: Yeah its ok.

Its complaining because we are using the same variable to update the value but we only do it once every taskDuration. If it was done more frequently we should have used atomic references.

File: indexing-service/src/test/resources/log4j2.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  ~ under the License.
+  -->
+
+<Configuration status="WARN">
+  <Loggers>
+    <Logger level="debug" name="org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner" additivity="false">
+      <AppenderRef ref="Console"/>
+    </Logger>
+  </Loggers>
+</Configuration>
```
Comment (suneet-s) [inline indexing-service/src/test/resources/log4j2.xml:27]: Is this change needed?

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 8]
```diff
@@ -2092,4 +2108,33 @@ protected abstract void possiblyResetDataSourceMetadata(
   protected abstract boolean isEndOffsetExclusive();
 
   protected abstract TypeReference<List<SequenceMetadata<PartitionIdType, SequenceOffsetType>>> getSequenceMetadataTypeReference();
+
+  private void refreshMinMaxMessageTime()
+  {
+    minMessageTime = minMessageTime.plusMinutes(ioConfig.getRefreshRejectionPeriodsInMinutes().intValue());
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:2114]: Let's add an info log here. This seems like it would be important to know when the min max times are refreshed

**Review by suneet-s — COMMENTED** [review]
Thanks for this change. I have one last small ask to add some logging. Also, the docs in https://github.com/apache/druid/blob/master/docs/ingestion/supervisor.md#io-configuration mention that there are problems if the message rejection period is set too low - can that caveat be removed now?

The CI failure seems like it's ok to overrule since it is complaining about branch coverage about moved code and the code that is missing tests is debug logging.

File: indexing-service/src/test/resources/log4j2.xml (hunk 7 above)
Comment (adithyachakilam) [reply indexing-service/src/test/resources/log4j2.xml:27]: It atleast works locally for coverage.

**Comment by suneet-s** [comment]
Overruling CI. I had to re-trigger a bunch of failed ITs, and now they have all passed. The only remaining failures are code coverage.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 9]
```diff
@@ -267,6 +271,18 @@ public SeekableStreamIndexTaskRunner(
     this.ingestionState = IngestionState.NOT_STARTED;
     this.lockGranularityToUse = lockGranularityToUse;
 
+    minMessageTime = ioConfig.getMinimumMessageTime().or(DateTimes.MIN);
+    maxMessageTime = ioConfig.getMaximumMessageTime().or(DateTimes.MAX);
+
+    if (ioConfig.getRefreshRejectionPeriodsInMinutes() != null) {
+      Execs.scheduledSingleThreaded("RejectionPeriodUpdater-Exec--%d")
```
Comment (cryptoe) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: Who closes this exec service ?

Comment (cryptoe) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: This seems to be creating on thread for each supervisor. Seems a bit much no ? Can we use a common exec service for this ?

Comment (cryptoe) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: cc @adithyachakilam

Comment (adithyachakilam) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: https://github.com/apache/druid/pull/17490

Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: > This seems to be creating on thread for each supervisor. Seems a bit much no ? Can we use a common exec service for this ?

+1

Comment (adithyachakilam) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:278]: this is creating thread on task not on supervisor ?
